pkg/controller/node/node_controller.go

```diff
@@ -959,8 +960,11 @@ func (ctrl *Controller) updateCandidateMachines(pool *mcfgv1.MachineConfigPool,
 		ctrl.logPool(pool, "filtered to %d candidate nodes for update, capacity: %d", len(candidates), capacity)
 	}
 	if capacity < uint(len(candidates)) {
-		// Arbitrarily pick the first N candidates; no attempt at sorting.
-		// Perhaps later we allow admins to weight somehow, or do something more intelligent.
+		// when list is longer than maxUnavailable, rollout nodes in zone order, zones without zone label
+		// are done last from oldest to youngest. this reduces likelihood of randomly picking nodes
+		// across multiple zones that run the same types of pods resulting in an outage in HA clusters
```

Comment:
Hmm so if `maxUnavailable` > 1 then there can still be a point where we are draining nodes in both zone A and B, right?

For example given maxUnavailable = 2, and 3 nodes in zones `A` and `B`, we will take the pairs: `[(A1, A2), (A3, B1), (B2, B3)]`.

I think that will ordinarily be OK assuming that `maxUnavailable < nodesInZone` or so - even though we're "straddling" an upgrade across A and B, a HA pod on the middle will still be able to reschedule (if necessary) from A3 to A1/A2 and/or B1 to B2/B3.

But in the case where maxUnavailable >= sizeof(zone), we could still hit this condition.  I am not sure how realistic that is though. 

On the other hand, we could also constrain maxUnavailable to not be able to upgrade an entire zone at once pretty easily - I think that would be an even stronger statement about availability that would only slightly slow down upgrades in that case.

Reply:
IMO, I think part of this is letting users make their own choice and calibrate maxUnavailable properly. On our end,  we're _mitigating_  risk without slowing down the rollouts but not taking full responsibility or claiming 0 disruption.

Reply:
Honestly, if we wanted to be more specific, I think it should be via a pre-flight check (which don't exist but since this is about a user understanding/constructing a strong HA cluster that's where it really belongs).

Reply:
I'm OK with allowing them to bump maxUnavailable past (zonesize-1).  But it seems like a bit of a potential footgun.   Perhaps at least worth a message in the logs if this happens?

There is a definite tradeoff between availability and speed here - sometimes indeed users will want to sacrifice availability to patch faster (think critical kernel CVE).  So I guess if we did default to capping to (zone-1) we'd need a knob to override that.

But on the flip side the admin can forget they bumped maxUnavailable and be at risk of HA loss across later upgrades.  

(Hmm, maybe we should have something like maxUnavailableForThisUpgrade that auto-resets once an upgrade is complete)

Soo....anyways, I guess fine as is, with an optional logging/testing of this case.

Reply:
`on the flip side the admin can forget they bumped maxUnavailable and be at risk of HA loss across later upgrades.` Yeah I really wish we had preflight checks to reconfirm these types of settings and remind people in the moment of what their existing settings are with a clear option to change..

+1 will def add some logging

Reply:
After talking to Colin about this, logging here won't be correct as it is called potentially multiple times on different candidate list. This is potentially info that might belong in a preflight check allowing users to know the zone distritbution/max unavailable before initiating an upgrade giving them a full and accurate view of their pool. There is some future pdb work in the pipeline that may add a preflight check/ui updates and if so, I will also add maxUnavailable there for clarity.

This is however a not a blocker, so in the interest of velocity and since this PR makes upgrades in HA clusters less distruptive, lifting the WIP to merge.

There will also be a docs update to note this, I will let docs know to add a note that max unavailable should probably be <= your smallest zone if you want the least chance of disruption and link to best HA practices if we have something